### PR TITLE
fix(explore): update overwrite button on perm change

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
Currently the Explore view does not update the "can_overwrite" permission when the user updates the chart ownership, as it's populated from bootstrap data (rendered server-side when page is loaded). In addition, the "Save" button in the chart metadata modal is enabled when the owners array is initially empty, but after adding and removing a user the save button gets disabled.

This removes the `can_overwrite` property from the Explore bootstrap data and replaces it with state from Redux, making it possible to change the disabled state of the "Save (Overwrite)" button in the Save modal. In addition, the `slice.owners` property in bootstrap data is changed to provide numeric ids instead of "firstname lastname" (most other models were already returning owners in this format; it appears the "firstname lastname" results weren't being used anywhere in Explore). Also update some types and remove redundant `user_id` property from bootstrap data.

### AFTER
https://user-images.githubusercontent.com/33317356/130752708-1ed4a830-f2d6-487d-a3d4-10a1fe283539.mp4

### BEFORE
https://user-images.githubusercontent.com/33317356/130752813-aeaa593c-94e2-48a8-9705-397048386cc2.mp4

### TESTING INSTRUCTIONS
1. Login as Admin
2. Open the Charts page and edit the metadata of an example dataset that doesn't have any owners
3. Add and remove yourself as owner and notice how the Save button goes from being enabled to disabled without the owners changing. Click cancel.
4. Open an example chart in Explore that doesn't have any owners
5. Add yourself as an owner
6. Try to overwrite the chart - notice that the "Save (Overwrite)" radio button is disabled

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/explore/reducers/exploreReducer.js [hunk 1]
```diff
@@ -219,6 +219,7 @@ export default function exploreReducer(state = {}, action) {
         slice: {
           ...state.slice,
           ...action.slice,
+          owners: action.slice.owners ?? null,
```
Comment (villebro) [inline superset-frontend/src/explore/reducers/exploreReducer.js:222]: This had to be explicitly added, as the action is missing the `owners` prop when all owners are missing

**Comment by stephenLYZ** [comment]
Frontend looks good to me! Thanks for the fixing. Is there any other state that can be maintained on the frontend only?

**Review by zhaoyongjie — APPROVED** [review]
LGTM

**Comment by villebro** [comment]
@stephenLYZ :
> Frontend looks good to me! Thanks for the fixing. Is there any other state that can be maintained on the frontend only?

Good question; I did briefly look into moving a few other properties to the frontend, but the obvious ones in this endpoint (`can_add` and `can_download`) seemed like they were actually better handled in the backend. But I will definitely be on the lookout for similar simplifications in subsequent PRs.

**Pull request merged**